cmd/ed25519/ed25519.go

```diff
@@ -0,0 +1,105 @@
+package main
+
+import (
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"chain/crypto/ed25519"
+)
+
+func main() {
+	if len(os.Args) < 2 {
+		usage()
+	}
+	switch os.Args[1] {
+	case "gen":
+		_, prv, err := ed25519.GenerateKey(nil)
+		must(err)
+		os.Stdout.Write([]byte(prv))
+
+	case "pub":
+		prv, err := ioutil.ReadAll(os.Stdin)
+		must(err)
+		pub := ed25519.PrivateKey(prv).Public().(ed25519.PublicKey)
+		os.Stdout.Write([]byte(pub))
+
+	case "sign":
+		if len(os.Args) < 3 {
+			usage()
+		}
+		prvhex := os.Args[2]
+		prv, err := hex.DecodeString(prvhex)
+		must(err)
+		msg, err := ioutil.ReadAll(os.Stdin)
+		must(err)
+		sig := ed25519.Sign(ed25519.PrivateKey(prv), msg)
+		os.Stdout.Write(sig)
+
+	case "verify":
+		args := os.Args[2:]
+		if len(args) < 1 {
+			usage()
+		}
+		var silent bool
+		if args[0] == "-s" {
+			silent = true
+			args = args[1:]
+		}
+		if len(args) < 2 {
+			usage()
+		}
+		pub, err := hex.DecodeString(args[0])
+		must(err)
+		sig, err := hex.DecodeString(args[1])
+		must(err)
+		msg, err := ioutil.ReadAll(os.Stdin)
+		must(err)
+		ok := ed25519.Verify(ed25519.PublicKey(pub), msg, sig)
+		var (
+			output string
+			w      io.Writer
+		)
+		if ok {
+			output = "OK"
+		} else {
+			output = "BAD"
+		}
+		if silent {
+			w = os.Stderr
+		} else {
+			w = os.Stdout
+		}
+		fmt.Fprintln(w, output)
+		if silent && !ok {
+			os.Exit(1)
```

Comment:
Yeah, in non-silent mode it's not failure successfully to detect a bad signature. Using this in a script and parsing its output, you want it to exit with status 0 to not affect the flow of the script.